### PR TITLE
SPLAT-770: ibmcloud disconnected install and custom endpoint coverage

### PR DESCRIPTION
**Pull request by jianlinliu**
1. ibmcloud disconnect install integration
2. tidy up some rehearse installer jobs
3. always save bastion info timely to ensure deprovision works
4. disable ibmcloud version check before login
5. add custom endpoint coverage on ibmcloud-ipi-fips-ep
6. update ibmcloud-ipi-proxy to remove `proxy-config-generate` which is only needed for private cluster.

### CODE REVIEW

**Comment by jianlinliu** [comment]
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.15-installer-rehearse-ibmcloud

**Comment by jianlinliu** [comment]
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.15-installer-rehearse-ibmcloud

**Comment by jianlinliu** [comment]
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.15-installer-rehearse-ibmcloud

**Comment by jianlinliu** [comment]
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.15-installer-rehearse-ibmcloud

**Comment by jianlinliu** [comment]
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.15-installer-rehearse-ibmcloud

**Comment by jianlinliu** [comment]
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.15-installer-rehearse-ibmcloud

**Comment by jianlinliu** [comment]
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.15-installer-rehearse-ibmcloud

**Comment by jianlinliu** [comment]
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.15-installer-rehearse-ibmcloud

**Comment by jianlinliu** [comment]
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.15-installer-rehearse-ibmcloud

**Comment by jianlinliu** [comment]
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.15-installer-rehearse-ibmcloud

**Comment by jianlinliu** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-nightly-ibmcloud-ipi-private-ep-fips-f14

**Comment by jianlinliu** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-nightly-ibmcloud-ipi-private-ep-fips-f14

**Comment by jianlinliu** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-nightly-ibmcloud-ipi-ep-fips-f14

**Comment by jianlinliu** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.14-amd64-stable-ibmcloud-ipi-proxy-private-f14

**Comment by jianlinliu** [comment]
cc @MayXuQQ to review in advance.

**Comment by jianlinliu** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-ibmcloud-ipi-disconnected-private-f14

**Comment by jianlinliu** [comment]
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.15-installer-rehearse-ibmcloud-disconnected

**Comment by MayXuQQ** [comment]
4.16.0-0.nightly-2024-01-03-090352 still failed by [OCPBUGS-25887](https://issues.redhat.com/browse/OCPBUGS-25887) [IBMCloud] cluster install fails nodes stuck in node.cloudprovider.kubernetes.io/uninitialized

File: ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.15__amd64-nightly.yaml [hunk 1]
```diff
@@ -1940,6 +1940,26 @@ tests:
     test:
     - chain: openshift-e2e-test-qe-destructive
     workflow: cucushift-installer-rehearse-ibmcloud-ipi-dedicatedhost-static
+- as: ibmcloud-ipi-disconnected-private-f14
+  cron: 17 12 12,26 * *
+  steps:
+    cluster_profile: ibmcloud-qe-2
+    env:
+      E2E_RUN_TAGS: '@amd64 and @ibmcloud-ipi and @disconnected'
+      TEST_FILTERS: ~ChkUpgrade&;~NonPreRelease&;~Serial&;~Disruptive&;~ConnectedOnly&;~HyperShiftMGMT&;~MicroShiftOnly&
```
[the remark this answers is not in the document]
Comment (jianlinliu) [reply ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.15__amd64-nightly.yaml:1949]: `~` means excludes the tests, and other disconnected install set them, I just copied them here.

File: ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.15__amd64-stable.yaml [hunk 2]
```diff
@@ -468,6 +468,13 @@ tests:
     test:
     - chain: openshift-upgrade-qe-test-arch-migration
     workflow: cucushift-installer-rehearse-gcp-ipi
+- as: ibmcloud-ipi-disconnected-private-f14
+  cron: 12 5 2,16 * *
+  steps:
+    cluster_profile: ibmcloud-qe-2
+    test:
+    - chain: cucushift-installer-check-cluster-health
```
[the remark this answers is not in the document]
Comment (jianlinliu) [reply ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.15__amd64-stable.yaml:476]: this is stable ci, no e2e testing here.

File: ci-operator/step-registry/cucushift/installer/check/ibmcloud/custom-endpoints/cucushift-installer-check-ibmcloud-custom-endpoints-commands.sh [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+if [ -f "${SHARED_DIR}/proxy-conf.sh" ] ; then
+    source "${SHARED_DIR}/proxy-conf.sh"
+fi
+
+ret=0
+if [ -n "$SERVICE_ENDPOINT_COS" ]; then
+    service_endpoint=$(get_service_endpoint "COS" $SERVICE_ENDPOINT_COS)
+    image_registry_pod=$(oc -n openshift-image-registry get pod -l docker-registry=default --no-headers | awk '{print $1}' | head -1)
+    output_from_pod=$(oc -n openshift-image-registry exec ${image_registry_pod} -- env | grep  "REGISTRY_STORAGE_S3_REGIONENDPOINT")
```
[the remark this answers is not in the document]
Comment (jianlinliu) [reply ci-operator/step-registry/cucushift/installer/check/ibmcloud/custom-endpoints/cucushift-installer-check-ibmcloud-custom-endpoints-commands.sh:39]: `REGISTRY_STORAGE_S3_REGIONENDPOINT` is the ENV var name, I am grepping the var from env output.

File: ci-operator/step-registry/cucushift/installer/rehearse/ibmcloud/ipi/disconnected/private/provision/cucushift-installer-rehearse-ibmcloud-ipi-disconnected-private-provision-chain.yaml [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+  - ref: ipi-install-times-collection
+  - ref: enable-qe-catalogsource-disconnected
+  - ref: mirror-images-tag-images
+  - ref: set-sample-operator-disconnected
+  - chain: cucushift-installer-check
+  - ref: cucushift-installer-check-ibmcloud
+  - ref: cucushift-installer-check-ibmcloud-private
+  env:
+  - name: BASE_DOMAIN
+    default: "ibmcloud.qe.devcluster.openshift.com"
```
[the remark this answers is not in the document]
Comment (jianlinliu) [reply ci-operator/step-registry/cucushift/installer/rehearse/ibmcloud/ipi/disconnected/private/provision/cucushift-installer-rehearse-ibmcloud-ipi-disconnected-private-provision-chain.yaml:25]: In disconnected install, need to registry mirror registry in public dns zone, there is no a public domain named in `private-ibmcloud.qe.devcluster.openshift.com`.

File: ci-operator/step-registry/ibmcloud/provision/bastionhost/ibmcloud-provision-bastionhost-commands.sh [hunk 5]
```diff
@@ -93,13 +106,79 @@ if [ X"${bastion_public_ip}" == X"" ] || [ X"${bastion_private_ip}" == X"" ] ; t
...
+        dns_zone_id=$(eval "${cmd}")
+        [[ -z "${dns_zone_id}" ]] && echo "ERROR: Did not find dns zone id per the output of '${cmd}'" && exit 3
+
+        # creating
+        run_command "ibmcloud dns resource-record-create ${dns_zone_id} -i ${IBMCLOUD_DNS_INSTANCE_NAME} --type A --name ${mirror_registry_host} --ipv4 ${bastion_private_ip}"
+
+        # post-check
+        dns_record_id=$(ibmcloud dns resource-records ${dns_zone_id} -i ${IBMCLOUD_DNS_INSTANCE_NAME} -o json | jq -r --arg z "${mirror_registry_dns}" '.resource_records[] | select(.name==$z) | .id')
+        [[ -z "${dns_record_id}" ]] && echo "ERROR: Did not find dns record id" && exit 3
+        echo "ibmcloud dns resource-record-delete ${dns_zone_id} ${dns_record_id} -i ${IBMCLOUD_DNS_INSTANCE_NAME} -f || ture" >>"${SHARED_DIR}/ibmcloud_remove_resources_by_cli.sh"
```
[the remark this answers is not in the document]
Comment (jianlinliu) [reply ci-operator/step-registry/ibmcloud/provision/bastionhost/ibmcloud-provision-bastionhost-commands.sh:143]: ibmcloud_remove_resources_by_cli.sh is not only clean up the created dns, also other created resource by cli, e.g: vpc endpoint gateway.

File: ci-operator/step-registry/cucushift/installer/rehearse/ibmcloud/ipi/disconnected/private/provision/cucushift-installer-rehearse-ibmcloud-ipi-disconnected-private-provision-chain.yaml [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+  - chain: ibmcloud-provision-bastionhost
+  - chain: mirror-images-by-oc-adm
+  - chain: ipi-conf-ibmcloud
+  - ref: ipi-conf-ibmcloud-provisionednetwork
+  - ref: ipi-conf-ibmcloud-custom-endpoints
+  - ref: ipi-conf-mirror
+  - ref: proxy-config-generate
+  - ref: openshift-cluster-bot-rbac
+  - ref: ipi-install-hosted-loki
+  - ref: ipi-install-install
```
[the remark this answers is not in the document]
Comment (jianlinliu) [reply ci-operator/step-registry/cucushift/installer/rehearse/ibmcloud/ipi/disconnected/private/provision/cucushift-installer-rehearse-ibmcloud-ipi-disconnected-private-provision-chain.yaml:15]: ipi-install chain includes `ipi-install-rbac`, which duplicates with the one in `mirror-images-by-oc-adm` chain.

File: ci-operator/step-registry/cucushift/installer/check/ibmcloud/custom-endpoints/cucushift-installer-check-ibmcloud-custom-endpoints-commands.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (MayXuQQ) [reply ci-operator/step-registry/cucushift/installer/check/ibmcloud/custom-endpoints/cucushift-installer-check-ibmcloud-custom-endpoints-commands.sh:39]: grep the env value or the env name?

File: ci-operator/step-registry/ibmcloud/provision/bastionhost/ibmcloud-provision-bastionhost-commands.sh (hunk 5 above)
[the remark this answers is not in the document]
Comment (MayXuQQ) [reply ci-operator/step-registry/ibmcloud/provision/bastionhost/ibmcloud-provision-bastionhost-commands.sh:143]: How about saving the resources in different steps in different files, and cleanup separately?

File: ci-operator/step-registry/cucushift/installer/check/ibmcloud/custom-endpoints/cucushift-installer-check-ibmcloud-custom-endpoints-commands.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (jianlinliu) [reply ci-operator/step-registry/cucushift/installer/check/ibmcloud/custom-endpoints/cucushift-installer-check-ibmcloud-custom-endpoints-commands.sh:39]: env name

File: ci-operator/step-registry/ibmcloud/provision/bastionhost/ibmcloud-provision-bastionhost-commands.sh (hunk 5 above)
[the remark this answers is not in the document]
Comment (jianlinliu) [reply ci-operator/step-registry/ibmcloud/provision/bastionhost/ibmcloud-provision-bastionhost-commands.sh:143]: I prefer to save the exact remove commands when they are being provisioned, when deprovision, only need to run the remove commands directly, I do not think they deserve different files in different steps, once that, when deprovision, also have to introduce different deprovision steps.

File: ci-operator/step-registry/cucushift/installer/check/ibmcloud/custom-endpoints/cucushift-installer-check-ibmcloud-custom-endpoints-commands.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (MayXuQQ) [reply ci-operator/step-registry/cucushift/installer/check/ibmcloud/custom-endpoints/cucushift-installer-check-ibmcloud-custom-endpoints-commands.sh:39]: $ echo ${output_from_pod}
REGISTRY_STORAGE_S3_REGIONENDPOINT=https://s3.direct.us-east.cloud-object-storage.appdomain.cloud

**Comment by MayXuQQ** [comment]
checked with registry.ci.openshift.org/ocp/release:4.15.0-0.nightly-2024-01-05-012756 PASS

**Comment by jianlinliu** [comment]
/pj-rehearse periodic-ci-openshift-verification-tests-master-installer-rehearse-4.15-installer-rehearse-ibmcloud-disconnected

**Comment by jianlinliu** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-nightly-ibmcloud-ipi-disconnected-private-f14

File: ci-operator/step-registry/cucushift/installer/rehearse/ibmcloud/ipi/with-ep-cos/provision/cucushift-installer-rehearse-ibmcloud-ipi-with-ep-cos-provision-chain.yaml [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+  - name: SERVICE_ENDPOINT_VPC
+    default: ""
+  - name: SERVICE_ENDPOINT_ResourceController
+    default: ""
+  - name: SERVICE_ENDPOINT_ResourceManager
+    default: ""
+  - name: SERVICE_ENDPOINT_DNSServices
+    default: ""
+  - name: SERVICE_ENDPOINT_COS
+    default: "DEFAULT_ENDPOINT"
```
Comment (gpei) [inline ci-operator/step-registry/cucushift/installer/rehearse/ibmcloud/ipi/with-ep-cos/provision/cucushift-installer-rehearse-ibmcloud-ipi-with-ep-cos-provision-chain.yaml:26]: I'm not sure if we want to rename the `"DEFAULT_ENDPOINT"` to `"DEFAULT_PRIVATE_ENDPOINT"` in those scripts, actually I'm a bit confused to see the "DEFAULT_ENDPOINT" word in the first, it sounds like it's meaning the default public  endpoint being used by a basic IPI install, but it should be the default regional private endpoint.

File: ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.15__amd64-nightly.yaml [hunk 8]
```diff
@@ -2032,26 +2052,26 @@ tests:
     test:
     - ref: cucushift-installer-check-cluster-health
     workflow: cucushift-installer-rehearse-ibmcloud-ipi-ovn-mtu-migrate
-- as: ibmcloud-ipi-private-fips-f14
-  cron: 27 10 13,28 * *
+- as: ibmcloud-ipi-ep-fips-f14
```
Comment (gpei) [inline ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.15__amd64-nightly.yaml:2055]: The "ep" in the job name may be a little unclear for people who is not familiar with the detailed workflow, considering it's still a short job name, how about using "custom-endpoint" directly, or even using "custom-cos-endpoint" as we're only overriding the COS endpoint.

File: ci-operator/step-registry/cucushift/installer/rehearse/ibmcloud/ipi/with-ep-cos/provision/cucushift-installer-rehearse-ibmcloud-ipi-with-ep-cos-provision-chain.yaml (hunk 7 above)
Comment (jianlinliu) [reply ci-operator/step-registry/cucushift/installer/rehearse/ibmcloud/ipi/with-ep-cos/provision/cucushift-installer-rehearse-ibmcloud-ipi-with-ep-cos-provision-chain.yaml:26]: The most of default endpoints (except COS is using direct endpoint) we defined indeed is private endpoints, but I am not sure if there would be some other change in the future, to changed them to other custom endpoints, another point is to keep consistent with aws custom endpoint with the same keyword.

Comment (gpei) [reply ci-operator/step-registry/cucushift/installer/rehearse/ibmcloud/ipi/with-ep-cos/provision/cucushift-installer-rehearse-ibmcloud-ipi-with-ep-cos-provision-chain.yaml:26]: OK, sounds like maybe `DEFAULT_CUSTOM_ENDPOINT` would be more appropriate. If AWS custom endpoint is already using this value, let's keep it as now then.

File: ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.15__amd64-nightly.yaml (hunk 8 above)
Comment (jianlinliu) [reply ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.15__amd64-nightly.yaml]: updated.

**Comment by jianlinliu** [comment]
cc @liangxia to review.

**Pull request merged**